### ISSUE TITLE
XLA Warmup for Multiple Instances and Batch Sizes

### ISSUE THREAD

**Issue by jeisinge** [triton-inference-server/server#1522]
**Is your feature request related to a problem? Please describe.**
We are using Triton with a TensorFlow model with a variable number of batch sizes of 1 to 500.  To improve throughput and latency, we increased the number of [model instances](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/optimization.html#model-instances) and enabled XLA.  The XLA compilation process increases our latency quite a bit for the first queries.  So, we attempted to use [model warmup](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/model_configuration.html#model-warmup).

There are two issues with the model warmup.  First, the configuration is very long to specify 500 warmup queries.  Second, and most importantly, it doesn't appear to warmup XLA for all of the instances - we still get large latencies for the first hour on our model deployment.

**Describe the solution you'd like**
It would be great to be able to provide one warmup configuration and multiple batch sizes and/or ranges that are run on _all_ instances.

Even better would be to be able to do this type of warmup offline - when the docker image is built.

Also, it might be nice to be able to query the service for information on what has been compiled to XLA.

**Describe alternatives you've considered**
We think TensorRT with its compilation might be a solution for us.  Unfortunately, TensorRT fails for our model --- probably something to do with TF Estimator / TF Feature Columns.

Currently, we have removed the model warmup from Triton and added it to a companion service.

**Additional context**

**Comment by deadeyegoodwin** [comment]
What happens if you use only 1 model instance? Does warmup seem correct then? Warmup should be done for all instances but perhaps there is a bug. I'm not sure if TF/XLA needs to see every possible batch size. You could also try just warming up with the largest batch size you want to support.

**Comment by jeisinge** [comment]
Interesting.  I am _not_ seeing warmup happening with one instance.  With one instance and 8 warmups of batch size 1-8, it appears that batch size 1 is warmed up.  However, batch size 2+ is not.

When setting the instance group to >1, it appears that batch size 1 is warmed up for all instance groups.   However, like with instance group = 1, no batch sizes > 1 are warmed up.

Also, I just noted that warmups only work when I provide the full model configuration inputs and outputs - it cannot use a generated configuration.

For XLA and batch sizes, I haven't investigated the code, but from the outside, it appears that XLA needs to be compiled for every batch size that we send up to a max value.  In particular, it appears our model maxes out around 20.  If we warm everything up to 20 externally, than any batch size like 1000 responds relatively quickly.  My guess is that internally, it is batching multiple 20 together plus a remainder.  Again, I haven't looked at the code.

**Comment by deadeyegoodwin** [comment]
That is unexpected behavior. We will need to take a deeper look into the warmup code to see if there is some unexpected interaction with XLA.

**Comment by GuanLuo** [comment]
@jeisinge I tried warm up a graphdef and a savedmodel from batch 1 to 8 with / without XLA, and I do see Triton reporting that all warmup batches are being run.

What Triton version were you using? Can you try to use the latest and verify again? If the problem have not resolved, can you shared your model config and verbose server log?

**Comment by jeisinge** [comment]
Agreed - I also see log messages that indicate something is happening.  However, for my model, it doesn't appear to warm up --- I still have the manually send request to warm up.  Once, I do that, the model performs well.

I tested on 20.03.0, originally, and, just now 20.03.1.

The model configuration is [config-warmup-triton.txt](https://github.com/NVIDIA/triton-inference-server/files/4808055/config-warmup-triton.txt)

I tested initially with and without warmup:
[triton_no_warmup.log](https://github.com/NVIDIA/triton-inference-server/files/4808057/triton_no_warmup.log) and [triton_warmup.log](https://github.com/NVIDIA/triton-inference-server/files/4808058/triton_warmup.log).
You should note that without warmup, the model takes more request time to finally warm up.

For verbose output, it has proprietary information.  I have tried to obsfuscate, but I still would rather not post publically.  I'll contact @GuanLuo privately.

**Comment by jeisinge** [comment]
Per e-mail, I am adding the details here:

<pre><font color="#4E9A06"><b>jacob@adelberts</b></font>:<font color="#3465A4"><b>~/Documents/tsp/triton</b></font>$ cat triton-verbose.log | grep armup
I0620 14:53:04.832661 1 backend.cc:306] Generating w<font color="#CC0000"><b>armup</b></font> sample data for &apos;batch size 1&apos;
I0620 14:53:04.833782 1 backend.cc:306] Generating w<font color="#CC0000"><b>armup</b></font> sample data for &apos;batch size 2&apos;
I0620 14:53:04.834851 1 backend.cc:306] Generating w<font color="#CC0000"><b>armup</b></font> sample data for &apos;batch size 3&apos;
I0620 14:53:04.836263 1 backend.cc:146] model &apos;jt_model&apos; instance 0 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 1&apos;
I0620 14:53:14.196935 1 backend.cc:146] model &apos;jt_model&apos; instance 0 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 2&apos;
I0620 14:53:14.825088 1 backend.cc:146] model &apos;jt_model&apos; instance 0 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 3&apos;
I0620 14:53:14.868597 1 backend.cc:146] model &apos;jt_model&apos; instance 1 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 1&apos;
I0620 14:53:22.523805 1 backend.cc:146] model &apos;jt_model&apos; instance 1 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 2&apos;
I0620 14:53:23.187262 1 backend.cc:146] model &apos;jt_model&apos; instance 1 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 3&apos;
I0620 14:53:23.231430 1 backend.cc:146] model &apos;jt_model&apos; instance 2 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 1&apos;
I0620 14:53:30.803000 1 backend.cc:146] model &apos;jt_model&apos; instance 2 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 2&apos;
I0620 14:53:31.459643 1 backend.cc:146] model &apos;jt_model&apos; instance 2 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 3&apos;
I0620 14:53:31.502445 1 backend.cc:146] model &apos;jt_model&apos; instance 3 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 1&apos;
I0620 14:53:38.994349 1 backend.cc:146] model &apos;jt_model&apos; instance 3 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 2&apos;
I0620 14:53:39.620969 1 backend.cc:146] model &apos;jt_model&apos; instance 3 is running w<font color="#CC0000"><b>armup</b></font> sample &apos;batch size 3&apos;
</pre>

<pre><font color="#4E9A06"><b>jacob@adelberts</b></font>:<font color="#3465A4"><b>~/Documents/tsp/triton</b></font>$ cat triton-verbose.log | grep Running
2020-06-20 14:53:01.299914: I tensorflow/cc/saved_model/loader.cc:153] <font color="#CC0000"><b>Running</b></font> initialization op on SavedModel bundle at path: /models/jt_model/1/model.savedmodel
2020-06-20 14:53:02.389575: I tensorflow/cc/saved_model/loader.cc:153] <font color="#CC0000"><b>Running</b></font> initialization op on SavedModel bundle at path: /models/jt_model/1/model.savedmodel
2020-06-20 14:53:03.493198: I tensorflow/cc/saved_model/loader.cc:153] <font color="#CC0000"><b>Running</b></font> initialization op on SavedModel bundle at path: /models/jt_model/1/model.savedmodel
2020-06-20 14:53:04.605227: I tensorflow/cc/saved_model/loader.cc:153] <font color="#CC0000"><b>Running</b></font> initialization op on SavedModel bundle at path: /models/jt_model/1/model.savedmodel
I0620 14:53:04.836276 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:53:14.197071 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 2 request payloads
I0620 14:53:14.825226 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 3 request payloads
I0620 14:53:14.868608 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:53:22.523948 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 2 request payloads
I0620 14:53:23.187393 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 3 request payloads
I0620 14:53:23.231441 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:53:30.803104 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 2 request payloads
I0620 14:53:31.459778 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 3 request payloads
I0620 14:53:31.502460 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloads
I0620 14:53:38.994465 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 2 request payloads
I0620 14:53:39.621108 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 3 request payloads
I0620 14:54:22.404282 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:22.410340 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloadsI0620 14:54:22.410373 1 grpc_server.cc:1007] New request handler for InferHandler, 12
I0620 14:54:22.413416 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:54:22.409776 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:23.810556 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloads
I0620 14:54:23.812761 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:23.814964 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloadsI0620 14:54:23.814678 1 infer_request.cc:154] add original input: [0x0x7fb164026490] request id: , model: jt_model, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 1, priority: 0, timeout (us): 0
I0620 14:54:23.819000 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:25.888028 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:25.889679 1 base_backend.cc:456] input &apos;jt_feature_y&apos; is GPU tensor: 0I0620 14:54:25.889714 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloads
I0620 14:54:25.891485 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:25.892771 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:54:30.055773 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloads
I0620 14:54:30.056925 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:54:30.059459 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:30.059391 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:39.113555 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:54:39.114384 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloads
I0620 14:54:39.117991 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:39.118605 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:40.924243 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloads
I0620 14:54:40.923502 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:54:40.928835 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:40.930845 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:45.373386 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloadsI0620 14:54:45.373393 1 infer_request.cc:154] add original input: [0x0x7fb1640187c0] request id: , model: jt_model, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 3, priority: 0, timeout (us): 0
I0620 14:54:45.376365 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:54:45.378700 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:45.381050 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:54.760569 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:54.763363 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloadsI0620 14:54:54.763373 1 grpc_server.cc:1014] Process for InferHandler, rpc_ok=1, 39 step START
I0620 14:54:54.764312 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloadsI0620 14:54:54.764320 1 base_backend.cc:456] input &apos;jt_feature_y&apos; is GPU tensor: 0
I0620 14:54:54.762908 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:56.651896 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_0_gpu0 with 1 request payloads
I0620 14:54:56.653920 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_2_gpu0 with 1 request payloads
I0620 14:54:56.657284 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_1_gpu0 with 1 request payloads
I0620 14:54:56.659542 1 base_backend.cc:644] <font color="#CC0000"><b>Running</b></font> jt_model_0_3_gpu0 with 1 request payloads
</pre>

**Comment by GuanLuo** [comment]
It appears that the warmup feature is sending the request with proper batch size to your model, the issue is less likely to be the warmup. What if you duplicate the warmup to send the batch 2, batch 3 multiple times? And after warmup, what is the time spent if you send your request with alternating batch size between 1 and 3 instead of always sending the same batch size?

**Comment by jeisinge** [comment]
I apologize for not getting back sooner.

>What if you duplicate the warmup to send the batch 2, batch 3 multiple times?

I will test the duplicate batch size late next week.

> And after warmup, what is the time spent if you send your request with alternating batch size between 1 and 3 instead of always sending the same batch size?

When I send a single request with batch size 1, I sometimes get a non-optimized.  In this case, latency is high.  It seems that this gets optimized in the background.  The next request sometimes hit one of these optimized instances and sometimes does not - in which case it will be optimized.  During this optimization process, if I hit this instance, the response is very slow.  Alternating between batch sizes does not affect anything --- just my the likelihood of getting an optimized/non-optimized instance.  Every batch size needs to be optimized for every instance. 

In other words, sending `bs1, wait 1 minute, bs2, wait 1 minute, bs 3, wait 1 minute, bs1, wait 1 minute, ...` is the same as `b1, wait 1 minute, bs 1, wait 1 minute, bs1, wait 1 minute,  bs 1, wait 1 minute, bs 2, wait 1 minute, ...`.

**Comment by YanDavKMS** [comment]
I'm also facing the same issue.
The model is being warmed up but the XLA compilation is being done on first request for each batch size being sent.

**Comment by dzier** [comment]
Closing due to lack of activity. Please re-open if issue persists on most recent release.
